NEW: Project-wide Actions

jamesmcgill opened the pull request:
### Description

In-built set of Actions which is installed (currently inside InputManager.asset) and is available via `InputSystem.actions`
New UITK based editor installed in Project Settings provides a UI for editing these Actions

### Changes made

- New UITK Actions Editor (existing IMGUI editor is still used for normal `.inputactions` asset files)
- New property `InputSystem.actions` backed by InputManager.asset and EditorBuildSettings for Players

### Manual Testing

[Testing Document](https://docs.google.com/document/d/1myFVqfXnsr8GcFnMf-OojhaLqe4dntRlzal5rZu1wQA/edit)

### Checklist

Before review:

- [ ] Changelog entry added.
    - Explains the change in `Changed`, `Fixed`, `Added` sections.
    - For API change contains an example snippet and/or migration example.
    - FogBugz ticket attached, example `([case %number%](https://issuetracker.unity3d.com/issues/...))`.
    - FogBugz is marked as "Resolved" with *next* release version correctly set.
- [ ] Tests added/changed, if applicable.
    - Functional tests `Area_CanDoX`, `Area_CanDoX_EvenIfYIsTheCase`, `Area_WhenIDoX_AndYHappens_ThisIsTheResult`.
    - Performance tests.
    - Integration tests.
- [ ] Docs for new/changed API's.
    - Xmldoc cross references are set correctly.
    - Added explanation how the API works.
    - Usage code examples added.
    - The manual is updated, if needed.

During merge:

- [ ] Commit message for squash-merge is prefixed with one of the list:
    - `NEW: ___`.
    - `FIX: ___`.
    - `DOCS: ___`.
    - `CHANGE: ___`.
    - `RELEASE: 1.1.0-preview.3`.

duckets committed:
DOCF-3823 - pre release docs

duckets committed:
Added stub doc to InputSystem.actions

duckets committed:
Added compilation constant to enable global actions during docs generation

duckets committed:
Fixed whitespace typo

jamesmcgill committed:
Add InputSystem.actions property

jamesmcgill committed:
Add ProjectWideActions Test Sample

jamesmcgill committed:
Rename conditonal compilation flag for project-wide actions

duckets committed:
Update PreReleaseNotes.md

tidied whitespace in code sample

duckets committed:
Update InputSystem.cs

removed duplicate comment on private member

jamesmcgill committed:
Merge branch 'develop' into project-wide-actions-phase1

lyndon-unity commented on Packages/com.unity.inputsystem/InputSystem/Editor/ProjectWideActions/ProjectWideActionsAsset.cs:
```diff
@@ -0,0 +1,99 @@
+
+            AssetDatabase.AddObjectToAsset(asset, s_AssetPath);
+
+            // Make sure all the elements in the asset have GUIDs and that they are indeed unique.
```
I'd be tempted to split this out into a support function taking the asset as a parameter.

lyndon-unity replied:
In part because I think a 'restore defaults' functionality may use this later
https://jira.unity3d.com/browse/ISX-1564

lyndon-unity commented on Packages/com.unity.inputsystem/InputSystem/Editor/ProjectWideActions/ProjectWideActionsAsset.cs:
```diff
@@ -0,0 +1,99 @@
+                }
+            }
+
+            // Create sub-asset for each action. This is so that users can select individual input actions from the asset when they're
```
Similar to the above chunk of code I think this could be a support function

lyndon-unity commented on Packages/com.unity.inputsystem/InputSystem/Editor/ProjectWideActions/ProjectWideActionsTemplate.inputactions:
```diff
@@ -0,0 +1,995 @@
+                {
+                    "name": "",
+                    "id": "e15ca19d-e649-4852-97d5-7fe8ccc44e94",
+                    "path": "<Gamepad>/buttonNorth",
```
I don't like this gamepad next/previous binding and am discussing with Marika to move to the same axis
<Gamepad>/dpad/right

lyndon-unity replied:
Done this

lyndon-unity commented on Packages/com.unity.inputsystem/InputSystem/Editor/ProjectWideActions/ProjectWideActionsTemplate.inputactions:
```diff
@@ -0,0 +1,995 @@
+                {
+                    "name": "",
+                    "id": "1534dc16-a6aa-499d-9c3a-22b47347b52a",
+                    "path": "<Keyboard>/2",
```
This key 2 seems more appropriate for Next
Am discussing with Marika to switch this to key 1

lyndon-unity replied:
Done this

lyndon-unity commented on Packages/com.unity.inputsystem/InputSystem/Editor/ProjectWideActions/ProjectWideActionsTemplate.inputactions:
```diff
@@ -0,0 +1,995 @@
+                {
+                    "name": "",
+                    "id": "cbac6039-9c09-46a1-b5f2-4e5124ccb5ed",
+                    "path": "<Keyboard>/1",
```
This key 1 seems more appropriate for Previous
Am discussing with Marika to switch this to key 2

lyndon-unity replied:
Done this

lyndon-unity commented on Packages/com.unity.inputsystem/InputSystem/Editor/ProjectWideActions/ProjectWideActionsTemplate.inputactions:
```diff
@@ -0,0 +1,995 @@
+                    "initialStateCheck": true
+                },
+                {
+                    "name": "Scroll",
```
UI team preferred original ScrollWheel term - considering switching back with Marika

lyndon-unity replied:
If changing the bindings below will next updating too

lyndon-unity commented on Packages/com.unity.inputsystem/InputSystem/Editor/ProjectWideActions/ProjectWideActionsTemplate.inputactions:
```diff
@@ -0,0 +1,995 @@
+                    "initialStateCheck": true
+                },
+                {
+                    "name": "PrimaryAction",
```
UI team preferred original Click term - considering switching back with Marika

lyndon-unity replied:
If changing the bindings below will next updating too

lyndon-unity commented on Packages/com.unity.inputsystem/InputSystem/Editor/ProjectWideActions/ProjectWideActionsTemplate.inputactions:
```diff
@@ -0,0 +1,995 @@
+                    "initialStateCheck": false
+                },
+                {
+                    "name": "TertiaryAction",
```
UI team preferred original MiddleClick term - considering switching back with Marika

lyndon-unity replied:
If changing the bindings below will next updating too

lyndon-unity commented on Packages/com.unity.inputsystem/InputSystem/Editor/ProjectWideActions/ProjectWideActionsTemplate.inputactions:
```diff
@@ -0,0 +1,995 @@
+                    "initialStateCheck": false
+                },
+                {
+                    "name": "SecondaryAction",
```
UI team preferred original RightClick term - considering switching back with Marika

lyndon-unity replied:
If changing the bindings below will next updating too

lyndon-unity replied:
Done this

lyndon-unity commented on Packages/com.unity.inputsystem/InputSystem/Editor/Settings/InputSettingsBuildProvider.cs:
```diff
@@ -9,21 +9,31 @@ namespace UnityEngine.InputSystem.Editor
-            if (InputSystem.settings == null)
-                return;
+            m_OriginalPreloadedAssets = PlayerSettings.GetPreloadedAssets();
+            var preloadedAssets = PlayerSettings.GetPreloadedAssets();
```
Could do this to avoid calling playersettings function twice as it may have higher cost.

 var preloadedAssets = m_OriginalPreloadedAssets;

lyndon-unity commented on Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/InputActionsEditorSettingsProvider.cs:
```diff
@@ -1,4 +1,4 @@
-#if UNITY_EDITOR && UNITY_INPUT_SYSTEM_UI_TK_ASSET_EDITOR // We use some UITK controls that are only available in 2022.2 or later (MultiColumnListView for example)
+#if UNITY_EDITOR && UNITY_INPUT_SYSTEM_PROJECT_WIDE_ACTIONS
```
Is it worth keeping the comment?

lyndon-unity commented on Packages/com.unity.inputsystem/InputSystem/InputSystem.cs:
```diff
@@ -3009,6 +3009,56 @@ internal static bool ShouldDrawWarningIconForBinding(string bindingPath)
+                        s_projectWideActions, true);
+                }
+                #else
+                s_projectWideActions = Resources.FindObjectsOfTypeAll<InputActionAsset>().FirstOrDefault();
```
Is this line valid - could it end up picking up some other input action asset that we don't expect ?

lyndon-unity replied:
Can the resources areas only contain the pre built one or could a user tuck some in there too ?
Would it be better to iterate over all and match by name.

lyndon-unity commented on Packages/com.unity.inputsystem/InputSystem/Editor/Settings/InputSettingsBuildProvider.cs:
```diff
@@ -9,21 +9,31 @@ namespace UnityEngine.InputSystem.Editor
-            // Hence we ignore adding temporary objects to preloaded assets.
-            if (!EditorUtility.IsPersistent(InputSystem.settings))
+#if UNITY_INPUT_SYSTEM_PROJECT_WIDE_ACTIONS
+            EditorBuildSettings.TryGetConfigObject(InputSettingsProvider.kEditorBuildSettingsActionsConfigKey,
```
Why do we reference config object here. We don't seem to use m_ProjectWideActions  for anything

lyndon-unity replied:
Oh wait InputSystem.actions also pulls from the configobject.
In which case - why don't we just use that reference directly ?

(This might be ok, but it confused me a little, perhaps a little comment here would help if it remains the same)

jamesmcgill committed:
Ensure UITK action editor appears in Project Settings

lyndon-unity reviewed (commented):
Looks ok overall (some questions in PR), I will check it out and run it too.

Some questions on bindings like we discussed in person. Am following up on those.

Lacks the Documentation .md file addition. Can you add that here too.

jamesmcgill committed:
Fix project settings editor not populating actions

jamesmcgill committed:
Fix for dupliate editors shown

jfreire-unity committed:
FIX: Global Actions UI flickers when user expands actions (#1738)

lyndon-unity commented on Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/InputActionsEditorWindow.cs:
```diff
@@ -204,24 +204,6 @@ private InputActionAsset GetAssetFromDatabase()
-
-        public static void SaveAsset(SerializedObject serializedAsset, InputActionsEditorWindow currentWindow = null)
-        {
-            if ((focusedWindow == null || focusedWindow is not InputActionsEditorWindow) && currentWindow == null)
```
Is it ok this check is removed - I think Joao added this to make sure not used on the Input Asset flow where IMGUI still used. Perhaps its not really needed?

jamesmcgill committed:
Fix for API validation test failure

jamesmcgill committed:
Fix ProjectWide Actions Enable/Disable flow for tests

jamesmcgill committed:
Prevent assertion which may happen if ProjectWide Actions are Enabled without any devices attached yet in tests

jamesmcgill committed:
Disable ProjectWide actions in some tests that assume there are no other actions

lyndon-unity committed:
Fixed issue with Input Actions UI becoming blank due to error (ISX-1584)

lyndon-unity committed:
Updated define for documentation generation to include the project wide actions API

lyndon-unity committed:
Hiding search field while not implemented (ISX-1582)

lyndon-unity committed:
Updated control bindings based on latest review

Next/Previous updated to be more intuative/same axis
Names for UI Click* restored to original
Added keyboard binding for Attack
Added some XR bindings

Some reordering too
E.g. so crouch/jump next to each other
This makes the diff slightly harder

jamesmcgill committed:
Fix tests not compiling on older editors

lyndon-unity committed:
Ran formatting script to fix warnings

lyndon-unity committed:
Merge branch 'project-wide-actions-phase1' of github.com:Unity-Technologies/InputSystem into project-wide-actions-phase1

lyndon-unity committed:
Hiding save, auto save buttons in project wide input settings as its auto saving always atm (ISX-1556)

lyndon-unity committed:
Fixed delete key mapping to be 'delete' rather than 'space' (ISX-1552)

jamesmcgill committed:
Fix ActionState and StateBuffer going out of sync when entering tests

jamesmcgill committed:
Disable ProjectWide actions in some tests that assume there are no other actions or controls

jamesmcgill committed:
Temporarily disable some tests which have issues when ProjectWide Actions are enabled

jamesmcgill committed:
fix formatting

jamesmcgill committed:
Fix broken links

jamesmcgill committed:
improve api docs for InputSystem.actions

Pauliusd01 reviewed (approved):
Approving to be mergable to develop. Testing will be continued in develop

jfreire-unity reviewed (approved):
Approving so we can start testing.

jamesmcgill committed:
add changelog entry

jamesmcgill commented:
Merging in present state for pre-release testing
Known unrelated issues on trunk and 2023.2 with building player for jobs:
- mac_standalone_il2cpp 
- android il2cpp
